core/org.wso2.carbon.core/src/main/java/org/wso2/carbon/core/util/KeyStoreManager.java

```diff
@@ -437,6 +495,67 @@ public PrivateKey getDefaultPrivateKey() throws Exception {
                 "available only for the super tenant.");
     }
 
+    /**
+     * This method loads the private key of a given tenant key store
+     *
+     * @param keyStoreName name of the tenant key store
+     * @param alias        alias of the private key
+     * @return private key corresponding to the alias
+     */
+    public PrivateKey getTenantPrivateKey(String keyStoreName, String alias) throws Exception{
+
+        String path = RegistryResources.SecurityManagement.KEY_STORES + "/" + keyStoreName;
+        org.wso2.carbon.registry.api.Resource resource;
+        KeyStore keyStore;
+
+        if (registry.resourceExists(path)) {
+            resource = registry.get(path);
+        } else {
+            throw new SecurityException("Given Key store is not available in registry : " + keyStoreName);
+        }
+
+        CryptoUtil cryptoUtil = CryptoUtil.getDefaultCryptoUtil();
+        String encryptedPassword = resource
+                .getProperty(RegistryResources.SecurityManagement.PROP_PRIVATE_KEY_PASS);
+        String privateKeyPasswd = new String(cryptoUtil.base64DecodeAndDecrypt(encryptedPassword));
+
+        if (isCachedKeyStoreValid(keyStoreName)) {
+            keyStore = tenantKeyStores.get(keyStoreName).getKeyStore();
+            return (PrivateKey) keyStore.getKey(alias, privateKeyPasswd.toCharArray());
+        } else {
+            byte[] bytes = (byte[]) resource.getContent();
+            String keyStorePassword = new String(cryptoUtil.base64DecodeAndDecrypt(resource.getProperty(
+                    RegistryResources.SecurityManagement.PROP_PASSWORD)));
+            keyStore = KeyStore.getInstance(resource
+                    .getProperty(RegistryResources.SecurityManagement.PROP_TYPE));
+            ByteArrayInputStream stream = new ByteArrayInputStream(bytes);
+            keyStore.load(stream, keyStorePassword.toCharArray());
+
+            KeyStoreBean keyStoreBean = new KeyStoreBean(keyStore, resource.getLastModified());
+            updateKeyStoreCache(keyStoreName, keyStoreBean);
+            return (PrivateKey) keyStore.getKey(alias, privateKeyPasswd.toCharArray());
+        }
+    }
+
+    /**
+     * Get the private key, of a given custom keystore.
+     *
+     * @param keyStoreName name of the custom key store. Must include the prefix "CustomKeyStore/"
+     * @return Private key
+     * @throws Exception Carbon Exception for tenants other than tenant 0
+     */
+    public PrivateKey getCustomKeyStorePrivateKey(String keyStoreName) throws Exception {
+
+        OMElement config = KeyStoreUtil.getCustomKeyStoreConfig(keyStoreName, this.getServerConfigService());
```

Comment:
Let's add a validation to the keystore name to verify it's a customkeystore

Reply:
Fixed in https://github.com/wso2/carbon-kernel/commit/a0533c76776fd427f270477373ea6951ce2f147f